File: 프로그래머스/3/60063. 블록 이동하기/블록 이동하기.py
from collections import deque

def solution(board):
    n = len(board)
    goal = (n-1, n-1)
    
    # 로봇의 초기 상태: ((첫 번째 칸 좌표), (두 번째 칸 좌표))
    start = ((0, 0), (0, 1))
    que = deque([(start, 0)])  # 큐에 로봇 상태와 경과 시간을 저장
    visited = set([start])  # 방문한 위치와 상태를 저장
    
    # 네 방향 이동 (상, 하, 좌, 우)
    directions = [(-1, 0), (1, 0), (0, -1), (0, 1)]
    
    def is_valid(pos1, pos2):
        x1, y1 = pos1
        x2, y2 = pos2
        return 0 <= x1 < n and 0 <= y1 < n and 0 <= x2 < n and 0 <= y2 < n and board[x1][y1] == 0 and board[x2][y2] == 0
    
    while que:
        (pos1, pos2), time = que.popleft()
        if pos1 == goal or pos2 == goal:
            return time
        
        x1, y1 = pos1
        x2, y2 = pos2
        
        # 네 방향 이동
        for dx, dy in directions:
            next_pos1 = (x1 + dx, y1 + dy)
            next_pos2 = (x2 + dx, y2 + dy)
            if is_valid(next_pos1, next_pos2) and (next_pos1, next_pos2) not in visited:
                visited.add((next_pos1, next_pos2))
                que.append(((next_pos1, next_pos2), time + 1))
        
        # 회전
        if x1 == x2:  # 가로로 놓여있는 경우
            for i in [-1, 1]:  # 위쪽 또는 아래쪽으로 회전
                if 0 <= x1 + i < n and board[x1 + i][y1] == 0 and board[x2 + i][y2] == 0:
                    # 왼쪽 칸을 축으로 위로 회전
                    if is_valid((x1 + i, y1), (x1, y1)):
                        if ((x1 + i, y1), (x1, y1)) not in visited:
                            visited.add(((x1 + i, y1), (x1, y1)))
                            que.append((((x1 + i, y1), (x1, y1)), time + 1))
                    # 오른쪽 칸을 축으로 위로 회전
                    if is_valid((x2 + i, y2), (x2, y2)):
                        if ((x2 + i, y2), (x2, y2)) not in visited:
                            visited.add(((x2 + i, y2), (x2, y2)))
                            que.append((((x2 + i, y2), (x2, y2)), time + 1))
        elif y1 == y2:  # 세로로 놓여있는 경우
            for i in [-1, 1]:  # 왼쪽 또는 오른쪽으로 회전
                if 0 <= y1 + i < n and board[x1][y1 + i] == 0 and board[x2][y2 + i] == 0:
                    # 위쪽 칸을 축으로 왼쪽으로 회전
                    if is_valid((x1, y1 + i), (x1, y1)):
                        if ((x1, y1 + i), (x1, y1)) not in visited:
                            visited.add(((x1, y1 + i), (x1, y1)))
                            que.append((((x1, y1 + i), (x1, y1)), time + 1))
                    # 아래쪽 칸을 축으로 왼쪽으로 회전
                    if is_valid((x2, y2 + i), (x2, y2)):
                        if ((x2, y2 + i), (x2, y2)) not in visited:
                            visited.add(((x2, y2 + i), (x2, y2)))
                            que.append((((x2, y2 + i), (x2, y2)), time + 1))
    
    return -1

# Example usage:
# board = [
#     [0, 0, 0, 1, 1],
#     [0, 0, 0, 1, 0],
#     [0, 1, 0, 1, 1],
#     [1, 1, 0, 0, 1],
#     [0, 0, 0, 0, 0]
# ]
# print(solution(board))  # Expected output: 7

File: 프로그래머스/3/60063. 블록 이동하기/블록 이동하기.java
import java.util.*;

public class Solution {
    
    static class Position {
        int x1, y1, x2, y2, time;

        Position(int x1, int y1, int x2, int y2, int time) {
            this.x1 = x1;
            this.y1 = y1;
            this.x2 = x2;
            this.y2 = y2;
            this.time = time;
        }
    }

    public int solution(int[][] board) {
        int n = board.length;
        Position start = new Position(0, 0, 0, 1, 0);  // 로봇의 초기 위치
        Set<String> visited = new HashSet<>();  // 로봇이 방문했던 위치
        String initialPos = "0,0,0,1";  // 초기 위치를 문자열로 저장
        visited.add(initialPos);  // 방문처리

        int[] dx = {-1, 1, 0, 0};  // 상하좌우 이동
        int[] dy = {0, 0, -1, 1};  // 상하좌우 이동

        Queue<Position> queue = new LinkedList<>();
        queue.offer(start);

        while (!queue.isEmpty()) {
            Position pos = queue.poll();
            int x1 = pos.x1, y1 = pos.y1, x2 = pos.x2, y2 = pos.y2, time = pos.time;

            // 목표에 도달했을 때
            if ((x1 == n - 1 && y1 == n - 1) || (x2 == n - 1 && y2 == n - 1)) {
                return time;
            }

            // 상하좌우 이동
            for (int i = 0; i < 4; i++) {
                int nx1 = x1 + dx[i], ny1 = y1 + dy[i];
                int nx2 = x2 + dx[i], ny2 = y2 + dy[i];

                if (isValid(nx1, ny1, nx2, ny2, n, board) && !visited.contains(nx1 + "," + ny1 + "," + nx2 + "," + ny2)) {
                    visited.add(nx1 + "," + ny1 + "," + nx2 + "," + ny2);  // 방문처리
                    queue.offer(new Position(nx1, ny1, nx2, ny2, time + 1));
                }
            }

            // 회전
            if (x1 == x2) {  // 가로로 놓여있는 경우
                for (int i = -1; i <= 1; i += 2) {  // 위나 아래로 회전
                    if (isValid(x1 + i, y1, x2 + i, y2, n, board)) {
                        if (!visited.contains(x1 + "," + y1 + "," + (x1 + i) + "," + y1)) {
                            visited.add(x1 + "," + y1 + "," + (x1 + i) + "," + y1);
                            queue.offer(new Position(x1, y1, x1 + i, y1, time + 1));
                        }
                        if (!visited.contains(x2 + "," + y2 + "," + (x2 + i) + "," + y2)) {
                            visited.add(x2 + "," + y2 + "," + (x2 + i) + "," + y2);
                            queue.offer(new Position(x2, y2, x2 + i, y2, time + 1));
                        }
                    }
                }
            } else if (y1 == y2) {  // 세로로 놓여있는 경우
                for (int i = -1; i <= 1; i += 2) {  // 왼쪽이나 오른쪽으로 회전
                    if (isValid(x1, y1 + i, x2, y2 + i, n, board)) {
                        if (!visited.contains(x1 + "," + y1 + "," + x1 + "," + (y1 + i))) {
                            visited.add(x1 + "," + y1 + "," + x1 + "," + (y1 + i));
                            queue.offer(new Position(x1, y1, x1, y1 + i, time + 1));
                        }
                        if (!visited.contains(x2 + "," + y2 + "," + x2 + "," + (y2 + i))) {
                            visited.add(x2 + "," + y2 + "," + x2 + "," + (y2 + i));
                            queue.offer(new Position(x2, y2, x2, y2 + i, time + 1));
                        }
                    }
                }
            }
        }

        return -1;  // 목적지에 도달하지 못하면 -1 반환
    }

    // 두 위치가 유효한지 확인하는 함수
    private boolean isValid(int x1, int y1, int x2, int y2, int n, int[][] board) {
        return x1 >= 0 && y1 >= 0 && x2 >= 0 && y2 >= 0 && x1 < n && y1 < n && x2 < n && y2 < n
               && board[x1][y1] == 0 && board[x2][y2] == 0;
    }
}
